fix(cloudwatch): render region and accountId when directly set on metrics

TrevorBurnham opened the pull request:
### Issue #

Closes #28731

### Reason for this change

Currently, if a user creates a metric that includes `region` and `accountId`, those fields are omitted when the metric renders in a stack with those same values. The intended behavior is to omit those fields when they're implicitly set via `metric.attachTo(stack)`, not to omit them when set directly by the user.

### Description of changes

This is a second attempt at #29935, which was auto-closed after the pull request linter complained that there were no integration test changes. This time around I've tried to satisfy the linter. Otherwise, the changes are the same as before:

The key changes here are on the `Metric` class. Previously, `region` and `account` were public properties that were set by `metric.attachTo(stack)`, making it impossible to differentiate whether they were set directly or via `attachTo`. To differentiate them while maintaining backward compatibility, I took this approach:

1. `attachTo` sets internal properties called `stackRegion` and `stackAccount`. Setting `region` and `account` directly sets internal properties called `regionOverride` and `accountOverride`.
2. The public `region` and `account` properties are now getters that return the override (if set) and fall back on the stack properties.
3. The `toMetricConfig()` method returns the `region` and `account` from the getters, but also includes the `regionOverride` and `accountOverride`.

That way, everything that looks at `region` and `account` works the same way it did before, except in `metricGraphJson`, which skips the "if different from stack" tokenization if the overrides are set.

### Description of how you validated changes

1. Confirmed that all existing unit tests pass.
2. Added a new unit test to show that `region` and `account` are now preserved when set directly on a metric.
3. Modified an integration test to show how setting `region` and `account` on a metric affects the snapshot.

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation reviewed (dismissed):
The pull request linter has failed. See the aws-cdk-automation comment below for failure reasons. If you believe this pull request should receive an exemption, please comment and provide a justification.


A comment requesting an exemption should contain the text `Exemption Request`. Additionally, if clarification is needed add `Clarification Request` to a comment.

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ The first word of the pull request title should not be capitalized. If the title starts with a CDK construct, it should be in backticks "``".

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 2595f62c470e0bda92695c1ac2695820c057a011
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=14882fdd-92a4-4f6d-b2a4-7a0c9e48e7bf%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

TrevorBurnham commented:
I don't know what's going on in the [build log](https://ops-cdk-autobuild-prs-autobuildv2github-buildlogs-g9jsj1nfmnjp.s3.amazonaws.com/14882fdd-92a4-4f6d-b2a4-7a0c9e48e7bf/build.log?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=ASIAWYORDRBKHTB2SVXQ%2F20241128%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20241128T212305Z&X-Amz-Expires=600&X-Amz-SignedHeaders=host&X-Amz-Security-Token=IQoJb3JpZ2luX2VjEL7%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLWVhc3QtMSJGMEQCIHpwhgQxb4RDhV1%2B%2FAThoQuUN5klw%2BD6b3h%2B4tg3BoMjAiAfe%2FDiYjQYZKl44A35BMILFaezbhahmrJ8z6dXFC8%2FeirTAwhnEAMaDDQ2NDgzMTg3NTE1NiIMx4%2Fv%2BJxrsB2q6k33KrADoU%2Fy6%2BCJX%2B1yaG9gqL4cMRDViqB%2BtjaA8CNHbLHmJddKnPCpnpgbUBaRJ1pZSkqfge%2B%2BIymL42NBe8RYxhTTOgYp%2BgIPLx2GAJpie26uNOc7BhK2g6TD%2Bnr30zxdYPAFJO28XWd7340cPxg63IdlVzhCuOX9I5MV6BIktZ9Ckh%2Fx%2FphQxvA%2BM8tkD4LBDSl7xbdxAGrty7rcAL9U4vLIA5hxmMLyskLSaNesHZzA3jkLr0TzzhgmXdn4tk9WOvFuUkpcqZ347GpYJ%2F%2Bpr0myUCMli4QP9rrjkhMK6jYA9pQ%2F%2BReMEpAx4N4M7kRMkLMQf%2BqLIfOq7CSjT5oru5HeyUVSmCoi%2B0XL4YGXdStB7TSxBpiYfPPyrXcVa62fct1QoJud5PMur5gVkxrlDCLAnwCoo4ay%2B9YrQFDbaIKk5yvERL4sVLde%2Fdz2vlO8TmWCZS1IfCmOYv6YpxF6RymrrE44ev4jD8uBClM33FToR53EvnYc51dgYRdnTiNUe%2BK2gFfcQObFuft1TXvjs1Hq55d6onRUYuGXPliqZiddkXT%2BbIDeqoN6vP1hdpGNw5n0MLe%2Bo7oGOp8BpaYqZK6PLDKU91ezk6lwhvTYFiNVEy0mplYtNfM4POEfFW6R1tNOVt6vgW%2F%2BpPFQmm%2BY8L%2BWbtSG1wXStlljgIMdTR47NQKLWUfxHeg0aLWEdfmc5tS%2Fz9P7kuQRVcDZkzpTy73VAF0KKO9eO7zQHVbEoj5vE%2Bm%2Fx5QArbB9TQSUC1yiabezFfemKUVyLnKkEqqIuV5LY6LpGUbQlv3a&X-Amz-Signature=38eb3a81632c95a169dc3eff8c78c4a45bb7a02dc0c139c2c77d19449fe3cec2). It says:

```
Failed tasks:

- aws-cdk-lib:test
...
âŒ  Last command failed. Scroll up to see errors in log (search for '!!!!!!!!').
```

But there are no matches for `!!!!!!!!` anywhere else in the log. 😕

aws-cdk-automation commented:
<!--STALE PR-->
This PR has been in the BUILD FAILING state for 3 weeks, and looks abandoned. To keep this PR from being closed, please continue work on it. If not, it will automatically be closed in a week.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 91680a5e9f65f6dfa785a3d7adfddfae76bf96cf
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2dcce176-03a8-48c8-aa3a-a350b97906f5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

TrevorBurnham commented:
I tried again, but the build is still failing. It's very strange; the [build log](https://ops-cdk-autobuild-prs-autobuildv2github-buildlogs-g9jsj1nfmnjp.s3.amazonaws.com/2dcce176-03a8-48c8-aa3a-a350b97906f5/build.log?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=ASIAWYORDRBKCUT2D3LZ%2F20241221%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20241221T140121Z&X-Amz-Expires=600&X-Amz-SignedHeaders=host&X-Amz-Security-Token=IQoJb3JpZ2luX2VjEN7%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLWVhc3QtMSJHMEUCIFC8wy8UfJVmLF%2Fc1%2BvU3UiBeLs8TBCCj1%2F1AmkVKfzwAiEA0RA7zTKtZvaVNqa4hBl7A3z7kjvtTrcBoFRzJbaisEkq3AMIp%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FARADGgw0NjQ4MzE4NzUxNTYiDKo%2BYiJytyK5gbyF2SqwAz2U9TguXEvp8cyRWrm5E7WjiVNCvuzlu72039iVSrO6Um9v4AwcO%2BzdcCBvrU8luWT%2BjlxNikQGMkobrXpDUwE1CtEg%2Fg8ptwxyNPeYcdRn1DTpPImsAdl0d3xX8s%2BHIeXbQOJIbAkIbv0S%2Byz3a2e9jUOYPghMV%2F4KT%2B5S9b5cep3H9%2FrfAoxqEW4Yt0gNc2GqO1WXkcLj5c9SOx0jRGUzoyskHAV4o6zswffFuCn9FDScvFde%2FPNcavH8MXMMdjuhoW6l1IMjaqIE%2BfmtM4QtYh7lVVmm135rTh6STX1vAPuxpQGpfHbTmkgAvpEV3wW%2BxR%2F0KzoDMJWG2RIXIODrqASZVxmN52rdNfXY9rwP8KnqtO7Ry1lNnFdYXBVqj2uSET%2FEew%2FU%2BytmtKUYmmDyL9m3DnAAE6BjYGFQyAFvqOH0dqk8HeRXmz2lJvkWKgKVFscCpPeyIRfLKOVga0jAnpuap7cHCAjTSCSZQQi9YJ9AbstZPmrL4jEQfNmQFtyRC6V0VE1SOG6iPhJVUvqcdIn9IZ429%2FKTF%2FIKyX6sY4zlMfI28R%2F46B8tD%2F08JjCwlJu7BjqeAWdZFrgVSAyXaHPQNaBdhd514q0ffOnh75lPth5MdmcTAq203N%2FKyyl1G8pwYhSAtvbjAJG91nlp%2B1PNudJFqHVhnXxWESDvoI6FNviXrTCmK5BAPSFwE5UHQOxYm7kNQIKr9EsLU5ZtRgzcaXg%2FEhxIma6%2Fy5QfR6J4Lq8CxkO1vN5yMK3ncHQJ7JpdarIPa0XL%2Bw%2FnYgeiii8X3%2FS3&X-Amz-Signature=c0d8b27a39c97f719a5f5512fa3d2cae825114fa91459bd867dcd3b3e1116497) says:

```
Failed tasks:

- aws-cdk-lib:test
```

But the output for `aws-cdk-lib:test` is:

```
> aws-cdk-lib:test

@aws-cdk/region-info: [2K[1G@aws-cdk/region-info: [1myarn run v1.22.22[22m
aws-cdk-lib: [2K[1Gaws-cdk-lib: [1myarn run v1.22.22[22m
@aws-cdk/region-info: [2K[1G[2m$ yarn gen && cdk-build --skip-lint[22m
aws-cdk-lib: [2K[1G[2m$ jest[22m
@aws-cdk/region-info: [2K[1G@aws-cdk/region-info: [2m$ cdk-copy region-info[22m
@aws-cdk/region-info: Build times for @aws-cdk/region-info: Total time (7.7s) | /codebuild/output/src2453208760/src/github.com/aws/aws-cdk/node_modules/jsii/bin/jsii (7.5s) | cdk-copy region-info (0.1s)
@aws-cdk/region-info: [2K[1GDone in 8.24s.
```

gracelu0 commented:
@TrevorBurnham  Looking at the build logsI believe some unit tests need to be updated, e.g. 
```
aws-cdk-lib: FAIL aws-rds/test/cluster.test.ts (24.426 s)
aws-cdk-lib:   â— cluster â€º cluster supports metrics
aws-cdk-lib:     expect(received).toEqual(expected) // deep equality
aws-cdk-lib:     - Expected  - 4
aws-cdk-lib:     + Received  + 4
aws-cdk-lib:       Object {
aws-cdk-lib:     -   "account": "12345",
aws-cdk-lib:         "dimensions": Object {
aws-cdk-lib:           "DBClusterIdentifier": Object {
aws-cdk-lib:             "Ref": "DatabaseB269D8BB",
aws-cdk-lib:           },
aws-cdk-lib:         },
aws-cdk-lib:         "metricName": "CPUUtilization",
aws-cdk-lib:         "namespace": "AWS/RDS",
aws-cdk-lib:     -   "period": Duration {
aws-cdk-lib:     +   "period": Object {
aws-cdk-lib:           "amount": 5,
aws-cdk-lib:     -     "unit": TimeUnit {
aws-cdk-lib:     +     "unit": Object {
aws-cdk-lib:             "inMillis": 60000,
aws-cdk-lib:             "isoLabel": "M",
aws-cdk-lib:             "label": "minutes",
aws-cdk-lib:           },
aws-cdk-lib:         },
aws-cdk-lib:     -   "region": "us-test-1",
aws-cdk-lib:     +   "stackAccount": "12345",
aws-cdk-lib:     +   "stackRegion": "us-test-1",
aws-cdk-lib:         "statistic": "Average",
aws-cdk-lib:       }
aws-cdk-lib:       2732 |     });
aws-cdk-lib:       2733 |
aws-cdk-lib:     > 2734 |     expect(stack.resolve(cluster.metricCPUUtilization())).toEqual({
aws-cdk-lib:            |                                                           ^
aws-cdk-lib:       2735 |       dimensions: { DBClusterIdentifier: { Ref: 'DatabaseB269D8BB' } },
aws-cdk-lib:       2736 |       namespace: 'AWS/RDS',
aws-cdk-lib:       2737 |       metricName: 'CPUUtilization',
aws-cdk-lib:       at Object.<anonymous> (aws-rds/test/cluster.test.ts:2734:59)
```
You can search for `FAIL` to find the failing unit tests

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: dc876541f04c5ba6eaecb7451df36632727cb590
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9c54ecde-ed41-495f-8e82-8dbb7d865c2b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

TrevorBurnham commented:
@gracelu0 Thanks for pointing me in the right direction! I figured out the issue: The new getters were not enumerable, causing backward compatibility-breaking behavior that showed up in those unit tests. I've corrected this by explicitly making the getters enumerable using `Object.defineProperty`, and I've added a new unit test for that specific aspect of `Metric`'s behavior.

xazhao commented on packages/aws-cdk-lib/aws-cloudwatch/lib/metric.ts:
```diff
@@ -409,11 +440,25 @@ export class Metric implements IMetric {
     return this.with({
-      region: cdk.Token.isUnresolved(stack.region) ? undefined : stack.region,
-      account: cdk.Token.isUnresolved(stack.account) ? undefined : stack.account,
+      stackAccount: cdk.Token.isUnresolved(stack.account) ? undefined : stack.account,
+      stackRegion: cdk.Token.isUnresolved(stack.region) ? undefined : stack.region,
```
Why do we need to `stackAccount` and `stackRegion` here? Can we directly set `region` and `account`?

xazhao commented:
> making it impossible to differentiate whether they were set directly or via attachTo

Do we need to differentiate whether they're set directly or via attachTo?

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 2518dad19287c75598031a1e63f92f8a139d76c4
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=97f06f6f-fdef-4b4b-84fd-32a2e3edf338%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

TrevorBurnham commented:
> Do we need to differentiate whether they're set directly or via attachTo?

The intent is to fix https://github.com/aws/aws-cdk/issues/28731 while preserving as much of the existing behavior as possible. The background facts are:

* Currently, account/region values that match those of the stack are excluded from serialization, regardless of whether those values are set directly by the user or implicitly via `attachTo`.
* Excluding values set directly by the user is a bug.

That means there are two possible solutions:

1. Differentiate whether the account/region values were set directly or via `attachTo`, or
2. Always include account/region values in serialization, regardless of how they were set.

This PR takes the first approach in order to maximize backward compatibility: Users who aren't directly setting account/region values won't see any change in serialized output. The second approach would make for a simpler code change, but a large number of users would see their serialized output change.

xazhao commented:
> > Do we need to differentiate whether they're set directly or via attachTo?
> 
> The intent is to fix #28731 while preserving as much of the existing behavior as possible. The background facts are:
> 
> * Currently, account/region values that match those of the stack are excluded from serialization, regardless of whether those values are set directly by the user or implicitly via `attachTo`.
> * Excluding values set directly by the user is a bug.
> 
> That means there are two possible solutions:
> 
> 1. Differentiate whether the account/region values were set directly or via `attachTo`, or
> 2. Always include account/region values in serialization, regardless of how they were set.
> 
> This PR takes the first approach in order to maximize backward compatibility: Users who aren't directly setting account/region values won't see any change in serialized output. The second approach would make for a simpler code change, but a large number of users would see their serialized output change.

Got it. Thanks for the explanations. Approved.
